CDAP-13554 | CDAP-13982 | Pipelines with many sinks may exceed the Variable substi…

chawlaasumit opened the pull request:
CDAP-13982 | Pipelines with many sinks may exceed the Variable substitution depth in the Hadoop conf
CDAP-13554 | Malformed json exceptions in Mapreduce context config while deserializing plugins map

Commit:
CDAP-13554 | Pipelines with many sinks may exceed the Variable substitution depth in the Hadoop conf

Commit:
CDAP-13554 | CDAP-13982 | Test case added for var substitution issue

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -24,50 +23,107 @@
+
+    @Test
+    public void testManyMacrosInAppSpec() {
+        Configuration hConf = new Configuration();
```
Please don't change the indentation. Indentation should be 2 spaces.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -24,50 +23,107 @@
+  for (String alphabet : alphabetsArr) {
+     hConf.set(alphabet, alphabet);
   }
+  LinkedHashSet<ArtifactId> parents = new LinkedHashSet<>();
```
The type should be `Set`.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -24,50 +23,107 @@
+  LinkedHashSet<ArtifactId> parents = new LinkedHashSet<>();
+  Plugin filePlugin1 = new Plugin(parents, artifactId,
+     new PluginClass("type", "name", "desc", "clsname", "cfgfield",
+  ImmutableMap.<String, PluginPropertyField>of()),
```
Wrong alignment. Also use `Collections.emptyMap()` instead.

chtyim commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -24,50 +23,107 @@
+
+  Map<String, Plugin> plugins = cfg.getPlugins();
+  Assert.assertEquals(plugins.size(), 1);
+ }
```
Wrong indentation. Please use 2 spaces.

Commit:
CDAP-13554 | CDAP-13982 | Updated as per comments

nitinmotgi commented:
@chawlaasumit if you can address comments and in the meantime, we will send you the CCLA / ICLA to sign (One time process). A +1 and signed CCLA/ICLA is required to merge this PR.

chawlaasumit commented:
> @chawlaasumit if you can address comments and in the meantime, we will send you the CCLA / ICLA to sign (One time process). A +1 and signed CCLA/ICLA is required to merge this PR.

I addressed the comments in last commit. https://github.com/caskdata/cdap/pull/10759/commits/6a9ad6970ee0fceabd14eaa7125b41b6c102af64. Please check once. Thanks!

chawlaasumit commented:
I addressed the comments in last commit. 6a9ad69
<https://github.com/caskdata/cdap/commit/6a9ad6970ee0fceabd14eaa7125b41b6c102af64>.
Please check once. Thanks! Regards

On Sat, Oct 6, 2018 at 3:24 PM Joltie Root <notifications@github.com> wrote:

> @chawlaasumit <https://github.com/chawlaasumit> if you can address
> comments and in the meantime, we will send you the CCLA / ICLA to sign (One
> time process). A +1 and signed CCLA/ICLA is required to merge this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/caskdata/cdap/pull/10759#issuecomment-427577624>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ApohiwEZCcA3fVEYc8qSMiJL9D1nq-IQks5uiL0RgaJpZM4XEorN>
> .
>

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -24,24 +24,48 @@
+  private static final Type PLUGIN_MAP_TYPE = new TypeToken<Map<String, Plugin>>() {
+  }.getType();
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+    .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
```
doesn't look like any of these type adapters are actually needed

albertshau replied:
this still does not look like it's needed. Does it fail if you use a normal new Gson()?

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -51,23 +75,54 @@ public void testManyMacrosInAppSpec() {
+                                    PluginProperties.builder().addAll(properties).build());
+
+    mockPlugins.put("File1", filePlugin1);
+    hConf.set(HCONF_ATTR_PLUGINS, GSON.toJson(mockPlugins, PLUGIN_MAP_TYPE));
```
don't think PLUGIN_MAP_TYPE is needed. you can just call GSON.toJson(mockPlugins)

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -51,23 +75,54 @@ public void testManyMacrosInAppSpec() {
-      Collections.<String, WorkerSpecification>emptyMap(),
-      Collections.<String, Plugin>emptyMap()
-    );
+    ApplicationSpecification appSpec =
```
It looks like the changes to this method are just indentation. If so, please change it back so it's clear there are no changes for the reviewer.

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -24,24 +24,48 @@
+  private static final Gson GSON = ApplicationSpecificationAdapter.addTypeAdapters(new GsonBuilder())
+    .registerTypeAdapter(Arguments.class, new ArgumentsCodec())
+    .registerTypeAdapter(ProgramOptions.class, new ProgramOptionsCodec()).create();
+  private static final String HCONF_ATTR_PLUGINS = "cdap.mapreduce.plugins";
```
instead of duplicating this variable, make the one in MapReduceContextConfig package private and use that one. Otherwise if somebody decides to change it there, this test will fail.

albertshau commented on cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java:
```diff
@@ -51,23 +75,54 @@ public void testManyMacrosInAppSpec() {
+    hConf.set(HCONF_ATTR_PLUGINS, GSON.toJson(mockPlugins, PLUGIN_MAP_TYPE));
+
+    Map<String, Plugin> plugins = cfg.getPlugins();
+    Assert.assertEquals(plugins.size(), 1);
```
should assert that the plugins are the same, not just the size.

Assert.assertEquals(mockPlugins, plugins)

chawlaasumit replied:
mockPlugins and plugins are not equal as PluginProperties build method is initializing macros with HashSet but the return from the GSON.fromJson is LinkedHashSet.

albertshau replied:
Set equality does not depend on the Set implementation. Are you sure they are not different?

chawlaasumit replied:
In Plugin.java, only the macros equality check is failing.
&& Objects.equals(properties.getMacros(), that.properties.getMacros());
Created a simple test
Set<String> lh1 = new LinkedHashSet();
Set<String> h1 = new HashSet();
Macros m1 = new Macros(lh1);
Macros m2 = new Macros(h1);
System.out.println("Is m1 equal to m2: "+ m1.equals(m2));

Output
Is m1 equal to m2: false

albertshau replied:
This has nothing to do with different set implementations. It is because the Macros class does not implement equals() and hashCode() on the release/4.3 branch. They are there on develop, you can copy-paste from there.

Commit:
CDAP-13554 | CDAP-13982 | Updated as per review comments

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfig.java:
```diff
@@ -72,7 +72,7 @@
   private static final String HCONF_ATTR_PROGRAM_ID = "cdap.mapreduce.program.id";
   private static final String HCONF_ATTR_WORKFLOW_INFO = "cdap.mapreduce.workflow.info";
-  private static final String HCONF_ATTR_PLUGINS = "cdap.mapreduce.plugins";
+  public static final String HCONF_ATTR_PLUGINS = "cdap.mapreduce.plugins";
```
this should be package private, not public. In other words, remove 'public'.

Commit:
CDAP-13554 | CDAP-13982 | Updated as per review comments

albertshau reviewed (approved):
lgtm. @nitinmotgi did we get a signed CCLA/ICLA?

nitinmotgi commented:
@chawlaasumit - We would appreciate if you could either sign ICLA https://cla.developers.google.com/about/google-individual or CCLA https://cla.developers.google.com/about/google-corporate

This is only done the first time, after that we will be able to merge your approved PRs. 

Thanks again for your contributions.

chawlaasumit commented:
I signed this last week. Thanks!

On Sat, Oct 20, 2018 at 3:08 PM Joltie Root <notifications@github.com>
wrote:

> We would appreciate if you could either sign ICLA
> https://cla.developers.google.com/about/google-individual or CCLA
> https://cla.developers.google.com/about/google-corporate
>
> This is only done the first time, after that we will be able to merge your
> approved PRs.
>
> Thanks again for your contributions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/caskdata/cdap/pull/10759#issuecomment-431585042>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Apohi-lX7b_ak7Pc0mM9JkfwTYeX7B7wks5umy5cgaJpZM4XEorN>
> .
>
